clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java

```diff
@@ -314,6 +319,31 @@ private <T> T execute(
       Map<String, String> headers,
       Consumer<ErrorResponse> errorHandler,
       Consumer<Map<String, String>> responseHeaders) {
+
+    // beforeConnectHandle is a pre-connection handler that needs to be executed before the first
+    // HTTP request.
+    // During execution, beforeConnectHandle must be unloaded to prevent recursive calls.
+    // After beforeConnectHandle executes successfully, other requests can be processed normally.
+    // If an exception occurs during the execution of beforeConnectHandle,
+    // likely due to a network issue, beforeConnectHandle needs to be reloaded to ensure
+    // it can be successfully executed once the connection is restored.
+    // Double-checking is used to avoid unnecessary locking overhead after a successful connection
+    if (beforeConnectHandle != null) {
+      synchronized (this) {
+        if (beforeConnectHandle != null) {
+          Function<HTTPClient, Void> handle = beforeConnectHandle;
+          beforeConnectHandle = null;
+          try {
+            handle.apply(this);
+          } catch (Exception e) {
+            // The beforeConnectHandle will continue processing until it succeeds.
+            beforeConnectHandle = handle;
```

Comment:
The variable swap seems a little bit tricky and hard to understand, do we have a better way to simplify the current code?

Reply:
Like using a flag to demonstrate that we've already done the check?